test/e2e/apps/job.go

```diff
@@ -100,6 +100,75 @@ var _ = SIGDescribe("Job", func() {
 		framework.ExpectEqual(successes, completions, "expected %d successful job pods, but got  %d", completions, successes)
 	})
 
+	ginkgo.It("should allow to use the pod failure policy on exit code to fail the job early", func() {
+
+		// We fail the Job's pod only once to ensure the backoffLimit is not
+		// reached and thus the job is failed due to the pod failure policy
+		// with FailJob action.
+		// In order to ensure a Job's pod fails once before succeeding we force
+		// the Job's Pods to be scheduled to a single Node and use a hostPath
+		// volume to persist data across new Pods.
+		ginkgo.By("Looking for a node to schedule job pod")
+		node, err := e2enode.GetRandomReadySchedulableNode(f.ClientSet)
+		framework.ExpectNoError(err)
+
+		ginkgo.By("Creating a job")
+		job := e2ejob.NewTestJobOnNode("failOnce", "pod-failure-failjob", v1.RestartPolicyNever, parallelism, completions, nil, backoffLimit, node.Name)
+		job.Spec.PodFailurePolicy = &batchv1.PodFailurePolicy{
+			Rules: []batchv1.PodFailurePolicyRule{
+				{
+					Action: batchv1.PodFailurePolicyActionFailJob,
+					OnExitCodes: &batchv1.PodFailurePolicyOnExitCodesRequirement{
+						Operator: batchv1.PodFailurePolicyOnExitCodesOpIn,
+						Values:   []int32{1},
+					},
+				},
+			},
+		}
+		job, err = e2ejob.CreateJob(f.ClientSet, f.Namespace.Name, job)
+		framework.ExpectNoError(err, "failed to create job in namespace: %s", f.Namespace.Name)
+
+		ginkgo.By("Ensuring job fails")
+		err = e2ejob.WaitForJobFailed(f.ClientSet, f.Namespace.Name, job.Name)
+		framework.ExpectNoError(err, "failed to ensure job failure in namespace: %s", f.Namespace.Name)
+	})
+
+	ginkgo.It("should allow to use the pod failure policy to not count the failure towards the backoffLimit", func() {
```

Comment:
since e2e tests are expensive to run, I'm fine with just having one test (the one above).

The important thing we are testing is that actions by kubelet can be matched with a job policy.

@aojea to comment whether is worth reducing the number of E2E tests. More through scenarios should be done at integration level for the job controller (which we did already).

Reply:
These tests seems to be very light, it seems to run only in 4 sec

> should allow to use the pod failure policy on exit code to fail the job early

E2E test are not only for testing the feature itself, also to see how it integrates with the rest of the system.

An example is the bug with the kubelet pod lifecycle, it was reported because someone was running job and its nature made the bug to be more exacerbated when using Jobs.

I would keep them, as I said , is adding only 4 secs